src/pal/src/exception/signal.cpp

```diff
@@ -242,33 +237,61 @@ void SEHCleanupSignals()
 #if !HAVE_MACH_EXCEPTIONS
 /*++
 Function :
-    sigill_handler
+    invoke_previous_action
 
-    handle SIGILL signal (EXCEPTION_ILLEGAL_INSTRUCTION, others?)
+    synchronously invokes the previous action or aborts when that is not possible
 
 Parameters :
-    POSIX signal handler parameter list ("man sigaction" for details)
+    action  : previous sigaction struct
+    code    : signal code
+    siginfo : signal siginfo
+    context : signal context
+    signalRestarts: BOOL state : TRUE if the process will be signalled again
 
     (no return value)
 --*/
-static void sigill_handler(int code, siginfo_t *siginfo, void *context)
+static void invoke_previous_action(struct sigaction* action, int code, siginfo_t *siginfo, void *context, bool signalRestarts = true)
 {
-    if (PALIsInitialized())
+    _ASSERTE(action != NULL);
+
+    bool hasSigInfo = action->sa_flags & SA_SIGINFO;
+
+    if (!hasSigInfo && action->sa_handler == SIG_IGN)
```

Comment:
The code is now difficult to read with all the hasSigInfo in all the conditions. I've thought you'd just completely separate the siginfo / non-siginfo cases, like this:
```C++
    if (action->sa_flags & SA_SIGINFO)
    {
        _ASSERTE(action->sa_sigaction != NULL);
        action->sa_sigaction(code, siginfo, context);
    }
    else
    {
        if (action->sa_handler == SIG_IGN)
        {
            if (signalRestarts)
            {
                // This signal mustn't be ignored because it will be restarted.
                PROCAbort();
            }
            return;
        }
        else if (action->sa_handler == SIG_DFL)
        {
            if (signalRestarts)
            {
                // Restore the original and restart h/w exception.
                restore_signal(code, action);
            }
            else
            {
                // We can't invoke the original handler because returning from the
                // handler doesn't restart the exception.
                PROCAbort();
            }
        }
        else
        {
            _ASSERTE(action->sa_handler != NULL);
            action->sa_handler(code);
        }
    }
```